### PR TITLE
Support Refresh Token API

### PR DESCRIPTION
**Pull request by dhei**
This PR to address Issue #62, a basic implementation that calls the refresh with "X-ZUMO-AUTH" and sets the mobile service auth token. The Refresh API includes "X-ZUMO-FEATURES" HTTP header for internal telemetry purpose as well.

@shrishrirang @pragnagopa can you review this? Thanks.

### CODE REVIEW

File: sdk/src/MSClient.h [hunk 1]
```diff
@@ -138,6 +138,9 @@
 /// Logs out the current end user.
 -(void)logoutWithCompletion:(nullable MSClientLogoutBlock)completion;
 
+/// Refreshes access token with the identity provider for the logged in user.
+-(void)refreshUserWithCompletion:(nullable MSClientLoginBlock)completion;
```
Comment (shrishrirang) [inline sdk/src/MSClient.h]: >refreshUserWithCompletion [](start = 7, length = 25)

will refreshWithCompletion be a better name? consistent with login and logout..

File: sdk/src/MSClientConnection.h [hunk 2]
```diff
@@ -38,6 +39,13 @@
               client:(MSClient *)client
           completion:(MSResponseBlock)completion;
 
+// Initializes an |MSClientConnection| with the given client sends the given
+// request with SDK features.
```
Comment (shrishrirang) [inline sdk/src/MSClientConnection.h]: "with the given client **AND** sends the given .." or something alike.

File: sdk/src/MSClientConnection.m [hunk 3]
```diff
@@ -77,6 +78,22 @@ -(id) initWithRequest:(NSURLRequest *)request
     return self;
 }
```
Comment (shrishrirang) [inline sdk/src/MSClientConnection.m]: This should call the new initializer you have defined.. Read about designated initializer here: [https://developer.apple.com/library/ios/documentation/General/Conceptual/CocoaEncyclopedia/Initialization/Initialization.html#//apple_ref/doc/uid/TP40010810-CH6-SW3](https://developer.apple.com/library/ios/documentation/General/Conceptual/CocoaEncyclopedia/Initialization/Initialization.html#//apple_ref/doc/uid/TP40010810-CH6-SW3)

File: sdk/src/MSClientConnection.m [hunk 4]
```diff
@@ -212,6 +229,18 @@ +(void) invokeNextFilter:(NSArray<id<MSFilter>> *)filters
...
+                                withFeatures:(MSSDKFeatures *)features
+{
+    NSMutableURLRequest *mutableRequest = [MSClientConnection configureHeadersOnRequest:request withClient:client];
+
+    [mutableRequest setValue:[MSSDKFeatures httpHeaderForFeatures:features]
+                    forHTTPHeaderField:MSFeaturesHeaderName];
+    
+    return mutableRequest;
+}
+
```
Comment (shrishrirang) [inline sdk/src/MSClientConnection.m]: would it be better to have a single configureHeadersOnRequest method?

File: sdk/src/MSError.h [hunk 5]
```diff
@@ -192,6 +192,17 @@ extern NSString *const MSErrorPushResultKey;
...
+/// Indicates that the refresh user operation failed because the identity provider
+/// does not support refresh token or user is not logged in with sufficient permission
+#define MSRefreshBadRequest                     -1507
+
+/// Indicates that the refresh user operation failed because credentials are not valid
+#define MSRefreshUnauthorized                   -1508
+
+/// Indicates that the refresh user operation failed because refresh token was revoked
+/// or expired
+#define MSRefreshForbidden                      -1509
```
Comment (shrishrirang) [inline sdk/src/MSError.h]: Is this our choice of error values or are they just mapping 1:1 with the error codes defined by EA?

File: sdk/src/MSLogin.m [hunk 6]
```diff
@@ -178,6 +179,58 @@ -(void) loginWithProvider:(NSString *)provider
     }
 }
 
+-(void)refreshUserWithCompletion:(nullable MSClientLoginBlock)completion
+{
+    // Create the request
+    NSURL *URL = [self.client.loginHost URLByAppendingPathComponent:@".auth/refresh"];
+    NSMutableURLRequest *request = [NSMutableURLRequest requestWithURL:URL];
+    request.HTTPMethod = @"GET";
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: >  request.HTTPMethod = @"GET"; [](start = 2, length = 30)

isn't this the default?

File: sdk/src/MSLogin.m [hunk 7]
```diff
@@ -178,6 +179,58 @@ -(void) loginWithProvider:(NSString *)provider
...
+    
+    // Create the response completion block
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: >// Create the response completion block [](start = 4, length = 39)

[minor]: May be this should go one line below and read something like - "Define a response completion block if required"?

File: sdk/src/MSLogin.m [hunk 8]
```diff
@@ -178,6 +179,58 @@ -(void) loginWithProvider:(NSString *)provider
...
+                if (response.statusCode < 400) {
+                    user = [[MSLoginSerializer loginSerializer] userFromData:data orError:&responseError];
+                    if (user) {
+                        self.client.currentUser = user;
+                    }
+                }
+                else {
+                    switch (response.statusCode) {
+                        case 400:
+                            responseError = [self errorWithDescriptionKey:@"Refresh failed with a 400 Bad Request error. The identity provider does not support refresh, or the user is not logged in with sufficient permission."
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: >The identity provider does not support refresh, or the user is not logged in with sufficient permission. [](start = 121, length = 104)

Is the error string provided by EA? I am trying to understand if we should re-define the error strings here or if we can simple use what EA returns..

File: sdk/src/MSLogin.m [hunk 9]
```diff
@@ -178,6 +179,58 @@ -(void) loginWithProvider:(NSString *)provider
...
+    request.HTTPMethod = @"GET";
+    
+    // Create the response completion block
+    MSResponseBlock responseCompletion = nil;
+    if (completion) {
+        responseCompletion = ^(NSHTTPURLResponse *response, NSData *data, NSError *responseError)
+        {
+            MSUser *user = nil;
+            if (!responseError) {
+                if (response.statusCode < 400) {
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: >400 [](start = 42, length = 3)

assumption is anything less than 400 is a successful completion? what about 3xx?

File: sdk/src/MSLogin.m [hunk 10]
```diff
@@ -221,4 +274,19 @@ -(NSURLRequest *) requestForProvider:(NSString *)provider
     return request;
 }
 
+
+#pragma mark * Private NSError Generation Methods
+
+
+-(NSError *) errorWithDescriptionKey:(NSString *)descriptionKey
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: >errorWithDescriptionKey [](start = 13, length = 23)

Looks like this is repeating in multiple files..

File: sdk/test/MSLoginTests.m [hunk 11]
```diff
@@ -0,0 +1,137 @@
...
+    NSLog(@"%@ tearDown", self.name);
+}
+
+#pragma mark * Refresh User Tests
+
+- (void)testRefreshUser
+{
+    MSClient *client = [MSClient clientWithApplicationURLString:@"http://someURL.com"];
+
+    MSTestFilter *testFilter = [MSTestFilter testFilterWithStatusCode:200];
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: >MSTestFilter [](start = 4, length = 12)

Can we also check for the outgoing request when alternateLoginHost is undefined, defined with trailing slash, defined without trailing slash, etc?

File: sdk/test/MSLoginTests.m [hunk 12]
```diff
@@ -0,0 +1,137 @@
...
+    
+    testFilter.onInspectResponseData = ^(NSURLRequest *request, NSData *data) {
+        NSDictionary *item = @{ @"user" : @{ @"userId" : @"sid:12345678" }, @"authenticationToken" : @"token12345678" };
+        return [[MSJSONSerializer JSONSerializer] dataFromItem:item idAllowed:YES ensureDictionary:NO removeSystemProperties:YES orError:nil];
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: >return [[MSJSONSerializer JSONSerializer] dataFromItem: [](start = 8, length = 55)

may be we should test that the client behavior is request when server returns an invalid (empty / malformed / missing fields like userId, etc) user..

File: sdk/test/MSLoginTests.m [hunk 13]
```diff
@@ -0,0 +1,137 @@
...
+- (void)testRefreshUser400Error
+{
+    MSClient *client = [MSClient clientWithApplicationURLString:@"http://someURL.com"];
+    
+    MSTestFilter *testFilter = [MSTestFilter testFilterWithStatusCode:400];
+    
+    MSClient *filterClient = [client clientWithFilter:testFilter];
+    
+    // Invoke the API
+    [filterClient refreshUserWithCompletion:
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: >refreshUserWithCompletion [](start = 18, length = 25)

can we add a test to check refresh works fine even if completion is nil?

File: sdk/test/MSLoginTests.m [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+     ^(MSUser *user, NSError *error) {
+         XCTAssertNotNil(error);
+         XCTAssertEqual(error.code, MSRefreshUnauthorized);
+         done = YES;
+     }];
+    
+    XCTAssertTrue([self waitForTest:1], @"Test timed out.");
+}
+
+- (void)testRefreshUser403Error
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: >testRefreshUser403Error [](start = 8, length = 23)

may be a test for error code outside of what you are checking for..

File: sdk/test/MSLoginTests.m [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+     }];
+    
+    XCTAssertTrue([self waitForTest:1], @"Test timed out.");
+}
+
+#pragma mark * Async Test Helper Method
+
+
+-(BOOL) waitForTest:(NSTimeInterval)testDuration {
+
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: We should now use [XCTestExpectation waitForExpectationsWithTimeout] instead..

File: sdk/src/MSClient.h (hunk 1 above)
Comment (dhei) [inline sdk/src/MSClient.h]: refreshUserWithCompletion makes more sense than refreshWithCompletion for this class, so let's keep it.

---
In reply to: [77883183](https://github.com/Azure/azure-mobile-apps-ios-client/pull/93#discussion_r77883183) [](ancestors = 77883183)

File: sdk/src/MSLogin.m (hunk 8 above)
Comment (dhei) [inline sdk/src/MSLogin.m]: Easy Auth backend doesn't provide much meaningful message on 400/401/403 errors, so we want to generate more meaningful error message in the iOS client explaining what happened and what to do. This is why we want to emit such error messages here.

---
In reply to: [77884888](https://github.com/Azure/azure-mobile-apps-ios-client/pull/93#discussion_r77884888) [](ancestors = 77884888)

File: sdk/src/MSLogin.m (hunk 10 above)
Comment (dhei) [inline sdk/src/MSLogin.m]: I will refactor this to reduce code duplication

---
In reply to: [77885331](https://github.com/Azure/azure-mobile-apps-ios-client/pull/93#discussion_r77885331) [](ancestors = 77885331)

File: sdk/src/MSClientConnection.m [hunk 16]
```diff
@@ -1,227 +1,241 @@
...
+        
+        if ([request HTTPBody] &&
+             ![request valueForHTTPHeaderField:contentTypeHeader]) {
+            // Set the content type header
+            [mutableRequest setValue:jsonContentType
+                  forHTTPHeaderField:contentTypeHeader];
+        }
+    }
+    
+    if (features) {
```
Comment (shrishrirang) [inline sdk/src/MSClientConnection.m]: >if (features) [](start = 4, length = 13)

The implicit assumption here is that MSFeatureNone is defined as 0. For example, when features == MSFeatureNone, the above condition will be true or false depending on the value of MSFeatureNone. Simply calling httpHeaderForFeatures: and relying on it to return appropriate header would be a better approach..

File: sdk/src/MSError.h [hunk 17]
```diff
@@ -192,6 +192,20 @@ extern NSString *const MSErrorPushResultKey;
 /// was invalid.
 #define MSLoginInvalidURL                       -1506
 
+/// Indicates that the refresh user operation failed because the identity provider
+/// does not support refresh token or user is not logged in with sufficient permission
+#define MSRefreshBadRequest                     -1507
```
Comment (shrishrirang) [inline sdk/src/MSError.h]: >-1507 [](start = 48, length = 5)

should we start at -1520 or something like that if we have to add some MSLogin* error codes in the future.. ?

File: sdk/src/MSLogin.m [hunk 18]
```diff
@@ -178,6 +180,63 @@ -(void) loginWithProvider:(NSString *)provider
...
+        // Define a response completion block if required
+        responseCompletion = ^(NSHTTPURLResponse *response, NSData *data, NSError *responseError)
+        {
+            MSUser *user = nil;
+            if (!responseError) {
+                if (response.statusCode == 200) {
+                    user = [[MSLoginSerializer loginSerializer] userFromData:data orError:&responseError];
+                    if (user.mobileServiceAuthenticationToken) {
+                        self.client.currentUser.mobileServiceAuthenticationToken = user.mobileServiceAuthenticationToken;
+                    }
```
Comment (shrishrirang) [inline sdk/src/MSLogin.m]: else Error?

File: sdk/test/MSLoginTests.m [hunk 19]
```diff
@@ -0,0 +1,175 @@
...
+#import <XCTest/XCTest.h>
+#import "MSClient.h"
+#import "MSJsonSerializer.h"
+#import "MSLogin.h"
+#import "MSTestFilter.h"
+#import "MSUser.h"
+
+@interface MSLoginTests : XCTestCase {
+    BOOL done;
```
Comment (shrishrirang) [inline sdk/test/MSLoginTests.m]: >BOOL done; [](start = 4, length = 10)

not needed anymore?

**Comment by shrishrirang** [comment]
:shipit:

File: sdk/src/MSClientConnection.m (hunk 16 above)
Comment (dhei) [inline sdk/src/MSClientConnection.m]: Good point, I will do that.

File: sdk/src/MSError.h (hunk 17 above)
Comment (dhei) [inline sdk/src/MSError.h]: Yes. Originally I thought Refresh API is part of Authorization / Authentication flow so I name following Login errors. Now I think it would be better to keep Login errors separate with Refresh errors.

---
In reply to: [78269757](https://github.com/Azure/azure-mobile-apps-ios-client/pull/93#discussion_r78269757) [](ancestors = 78269757)
